File: housePaint.py
'''
Time Compelxity -->
O(n)

Space Complexity -->
O(n) since number of columns is fixed

Approach -->
Go from bottom to top
make sure to keep the track of the index if suppose we at 1st column the for the next row choose min(0, 2)
and similarly for 2nd and the 0th column and keep updating the particular row and column with min val
'''

class Solution:
    def minCost(self, costs: List[List[int]]) -> int:
        
        #The 0, 1, 2 indicates R, B, G
        n = len(costs) 
        for i in range(n-2, -1, -1):
            for j in range(len(costs[i])):
                if j==0:
                    costs[i][j]+= min(costs[i+1][j+1], costs[i+1][j+2])
                elif j==1:
                    costs[i][j]+= min(costs[i+1][j-1], costs[i+1][j+1])
                else:
                    costs[i][j]+=min(costs[i+1][j-2], costs[i+1][j-1])
        print(costs)
        return min(costs[0])

File: house-paint.cpp
class Solution {
public:
    int minCost(vector<vector<int>>& costs) {
        int m = costs.size(); 
        int n = 3;
        int dp[m][n];
        dp[0][0] = costs[0][0];
        dp[0][1] = costs[0][1];
        dp[0][2] = costs[0][2];
        
        for(int i=1; i<m; i++){
            for(int j=0; j<n; j++){
                if(j==0){
                    dp[i][j] = min(dp[i-1][1], dp[i-1][2])+costs[i][j];
                }
                else if(j==1){
                    dp[i][j] = min(dp[i-1][0], dp[i-1][2])+costs[i][j];
                }
                else if(j==2){
                    dp[i][j] = min(dp[i-1][1], dp[i-1][0])+costs[i][j];
                }
            }
        }
        
        return min(min(dp[m-1][0], dp[m-1][1]), dp[m-1][2]);
    }
};

// TC: (3 * n) where n is the number of houses
// SC: (3 * n)
